public/app/plugins/panel/timeseries/TimeSeriesPanel.tsx

```diff
@@ -1,134 +1,221 @@
+import React, { useCallback, useMemo, useRef, useState } from 'react';
+import { DataFrame, EventBus, Field, PanelProps, TimeRange } from '@grafana/data';
+import {
+  UPlotChart2,
+  useTheme2,
+  VizLayout,
+  PlotLegend,
+  LegendDisplayMode,
+  ZoomPlugin,
+  TooltipPlugin,
+  usePanelContext,
+} from '@grafana/ui';
 import { getFieldLinksForExplore } from 'app/features/explore/utils/links';
-import { AnnotationsPlugin } from './plugins/AnnotationsPlugin';
-import { ContextMenuPlugin } from './plugins/ContextMenuPlugin';
-import { ExemplarsPlugin } from './plugins/ExemplarsPlugin';
 import { TimeSeriesOptions } from './types';
 import { prepareGraphableFields } from './utils';
-import { AnnotationEditorPlugin } from './plugins/AnnotationEditorPlugin';
-import { ThresholdControlsPlugin } from './plugins/ThresholdControlsPlugin';
 import { config } from 'app/core/config';
 import { PanelDataErrorView } from '@grafana/runtime';
+import { preparePlotConfigBuilder } from '@grafana/ui/src/components/TimeSeries/utils';
+import { debugLog } from '@grafana/ui/src/components/uPlotChart/debug';
+import { AnnotationsPlugin } from './plugins/AnnotationsPlugin';
+import { AnnotationEditorPlugin } from './plugins/AnnotationEditorPlugin';
+import { ContextMenuPlugin } from './plugins/ContextMenuPlugin';
+import { ExemplarsPlugin } from './plugins/ExemplarsPlugin';
+import { ThresholdControlsPlugin } from './plugins/ThresholdControlsPlugin';
+import { preparePlotData } from '@grafana/ui/src/components/uPlot/utils';
+import { PrepDataFnResult } from '@grafana/ui/src/components/uPlot/config/UPlotConfigBuilder';
+import { AlignedData } from 'uplot';
 
 interface TimeSeriesPanelProps extends PanelProps<TimeSeriesOptions> {}
 
 export const TimeSeriesPanel: React.FC<TimeSeriesPanelProps> = ({
   data,
-  timeRange,
-  timeZone,
   width,
   height,
   options,
   fieldConfig,
   onChangeTimeRange,
   replaceVariables,
   id,
+  timeZone,
+  ...otherProps
 }) => {
   const { sync, canAddAnnotations, onThresholdsChange, canEditThresholds, onSplitOpen } = usePanelContext();
+  const theme = useTheme2();
+  const timeRange = useRef<TimeRange>();
+  const eventBus = useRef<EventBus>();
+  const series = useRef<DataFrame[]>();
+  const alignedFrame = useRef<DataFrame | null>(null);
+  const [dataErrors, setDataErrors] = useState<any>();
+
+  const enableAnnotationCreation = Boolean(canAddAnnotations && canAddAnnotations());
+
+  timeRange.current = otherProps.timeRange;
+  eventBus.current = otherProps.eventBus;
+  series.current = data.series;
+
+  // Responsible for data validation and preparation
+  const plotData = useMemo(() => {
+    debugLog('TimeSeriesPanel.plotData memo');
+    const result = prepareTimeSeriesPlotData(data.series);
+
+    if (result.error) {
+      setDataErrors(result.error);
+    }
+    // Storing aligned frame via ref to avoid unnecessary config invalidation
+    alignedFrame.current = result.alignedFrame;
```

Comment:
@leeoniya  - we need to pass that via ref. Otherwise, we will get unnecessary config invalidation if we make `plotData` a dependency to `plotConfig` memo